### PR TITLE
feat: add writeFile sandbox global with write_dirs allowlist

### PR DESCRIPTION
**Pull request by panghy**
## Summary

Adds a `writeFile(path, data, options?)` global to the relay's JS sandbox (`execute_tools`), so tool scripts can persist generated media/artifacts (images, audio, downloads) to disk — gated by an explicit `write_dirs` allowlist.

User-visible behavior:

- New optional `write_dirs` array under `[relay]` config: a list of absolute directory paths that sandbox scripts are allowed to write into. When empty or absent, every write is denied — writes are strictly opt-in.
- Inside `execute_tools` scripts, `writeFile(absPath, data, opts?)` supports `utf8` (default) and `base64` encodings, so binary tool output (images, audio) can reach disk.
- Writes are atomic (temp file + rename) and parent directories are auto-created, but only inside the matched allowlisted root.
- Out-of-allowlist writes fail with an actionable error naming `[relay] write_dirs` and the desktop Settings page.
- End-to-end coverage includes an image round-trip integration test (MCP server → writeFile → read back byte-identical) plus hot-reload coverage of the new config key.

## Safety properties

- **Absolute-path allowlist**: only absolute destinations inside directories explicitly listed in `write_dirs` are writable; relative/empty/NUL paths and `..` components are rejected outright.
- **Canonical-prefix enforcement**: the fully-resolved destination (canonicalized deepest existing ancestor + remaining components) must sit inside one of the canonical `write_dirs` roots, so symlink and normalization escapes cannot break out of an allowed root.
- **Per-run resource limits**: 32 MB per file, 64 files per run, 256 MB total per run — a runaway script cannot exhaust disk.

## Linear

END-21 — add writeFile so tool media can reach the filesystem.

## Commits

- feat: add write_dirs allowlist config and plumb write roots to SandboxState
- feat: add writeFile sandbox global with write_dirs allowlist enforcement
- feat: enforce per-run writeFile resource limits (32 MB/file, 64 files, 256 MB total)
- test: end-to-end image round-trip integration test for writeFile

### CODE REVIEW

**Review by panghy — COMMENTED** [review]
test-probe (will be superseded)

File: src/js_sandbox.rs [hunk 1]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+/// here sit inside the matched root. On failure the temp file is removed —
+/// no partial destination file is ever observable.
+fn write_atomic(dest: &Path, bytes: &[u8]) -> Result<(), String> {
+    let dir = dest
+        .parent()
+        .ok_or_else(|| format!("writeFile: '{}' has no parent directory", dest.display()))?;
+    let file_name = dest
+        .file_name()
+        .ok_or_else(|| format!("writeFile: '{}' does not name a file", dest.display()))?;
+    std::fs::create_dir_all(dir).map_err(|e| {
```
Comment (panghy) [inline src/js_sandbox.rs:1213]: **Non-blocking (TOCTOU hardening)**: there is a validate→write window — `resolve_write_path` canonicalizes and containment-checks, then `write_atomic` runs `create_dir_all` + temp write + rename. If a validated existing directory is swapped for a symlink pointing outside the root in that window, the write follows it. This is not exploitable from inside the sandbox (`writeFile` only ever creates regular files, so a script cannot plant symlinks; a hostile local process is outside the threat model), but a cheap belt-and-suspenders fix would be to pass the matched root into `write_atomic`, re-`canonicalize(dir)` after `create_dir_all`, and re-assert `starts_with(root)` before writing the temp file. Fine to defer.

File: src/js_sandbox.rs [hunk 2]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+
+    let dest_str = dest.to_string_lossy();
+    Ok(JsValue::from(boa_engine::js_string!(dest_str.as_ref())))
+}
+
+/// Validate `path_str` against the allowlist and return the canonical
+/// destination path. Rejections (empty/NUL path, relative path, `..`
+/// components, destinations outside every root — including via symlinks)
+/// come back as `Err(message)` before anything touches the filesystem.
+fn resolve_write_path(path_str: &str, write_roots: &[PathBuf]) -> Result<PathBuf, String> {
```
Comment (panghy) [inline src/js_sandbox.rs]: **Non-blocking edge case**: if a canonical root is deleted *after* config resolution, a write to `<root>/sub/x` still passes — the existing-ancestor walk climbs above the deleted root, and the re-appended resolved path still `starts_with(root)` — and `write_atomic`'s `create_dir_all` then **recreates the deleted root**. Still inside the allowed prefix, so no security issue, but it slightly contradicts the "directories are never auto-created" stance in `resolve_write_roots`' docs (which technically only covers resolution time). Worth a one-line doc note here if intentional.

File: src/js_sandbox.rs [hunk 3]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+    let path_str = args
+        .first()
+        .ok_or_else(|| JsNativeError::typ().with_message("writeFile: missing path"))?
+        .to_string(context)?
+        .to_std_string_escaped();
+
+    let data = args
+        .get(1)
+        .ok_or_else(|| JsNativeError::typ().with_message("writeFile: missing data"))?
+        .to_string(context)?
```
Comment (panghy) [inline src/js_sandbox.rs]: **Nit**: `data` is coerced via `ToString`, so `writeFile(p, someObject)` silently writes `[object Object]` and `writeFile(p, [1,2,3])` writes `1,2,3`. Consider rejecting non-string `data` with a `TypeError` (`writeFile: data must be a string`) to surface script bugs early — media payloads are always strings here anyway.

File: src/js_sandbox.rs [hunk 4]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+            .ok_or_else(|| JsNativeError::error().with_message("sandbox state not initialised"))?;
+        Ok::<(Vec<PathBuf>, WriteLimits, usize, usize), JsError>((
+            state.write_roots.clone(),
+            state.write_limits,
+            state.files_written,
+            state.bytes_written,
+        ))
+    })?;
+
+    if files_written >= limits.max_files_per_run {
```
Comment (panghy) [inline src/js_sandbox.rs:1008]: **Nit**: `files_written` counts write *calls*, not distinct files — overwriting the same path 65 times trips "per-run limit of 64 files". The behavior seems right (it bounds I/O work), but consider wording the message as "64 file writes" so a script overwriting a single file understands why it was cut off.

File: src/js_sandbox.rs [hunk 5]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+                        ""
+                    },
+                    estimated_len,
+                    limits.max_file_bytes
+                ))
+                .into());
+        }
+    }
+
+    let bytes: Vec<u8> = match encoding.as_str() {
```
Comment (panghy) [inline src/js_sandbox.rs:1061]: **Nit (check ordering)**: path validation (`resolve_write_path`) runs after base64 decoding, so a write to a disallowed path still pays up to a 32 MB decode allocation, and a script with both a bad path and a bad payload sees the encoding error instead of the more actionable path error. Calling `resolve_write_path` before the decode would fail faster and improve error precedence. Purely efficiency/UX.

File: src/js_sandbox.rs [hunk 6]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+
+    if !write_roots.iter().any(|root| resolved.starts_with(root)) {
+        return Err(write_dirs_rejection_message(path_str, write_roots));
+    }
+    Ok(resolved)
+}
+
+/// The actionable "not inside a configured write directory" message. When
+/// roots are configured, the currently-allowed directories are listed.
+fn write_dirs_rejection_message(path_str: &str, write_roots: &[PathBuf]) -> String {
```
Comment (panghy) [inline src/js_sandbox.rs:1175]: **Nit (message consistency)**: the path is interpolated unquoted here (`{} is not inside...`) while every other `writeFile` message quotes it as `'{}'`. Quoting helps when the path has leading/trailing whitespace or looks like prose.

File: src/watcher.rs [hunk 7]
```diff
@@ -222,6 +228,26 @@ async fn reload_and_apply(
...
+    // changed. Every MetaToolHandler (global and per-profile) holds this
+    // same handle, so the next execute_tools snapshot sees the new roots.
+    let new_write_roots = config::resolve_write_roots(&new_config);
+    {
+        let changed = write_roots
+            .read()
+            .map(|current| *current != new_write_roots)
+            .unwrap_or(true);
+        if changed {
+            if let Ok(mut guard) = write_roots.write() {
```
Comment (panghy) [inline src/watcher.rs]: **Nit (poisoned-lock recovery)**: if the lock is ever poisoned, the `read().unwrap_or(true)` above correctly falls through to attempting the update, but `write()` also returns `Err` here, so the swap is silently skipped on every subsequent reload — the allowlist would be permanently frozen. `write().unwrap_or_else(|p| p.into_inner())` would recover (the guarded data is a plain `Vec<PathBuf>`, safe to reuse) and keep hot reload functional. This mirrors the poison-degrade choice already made on the read side in `MetaToolHandler::execute_tools`.

**Review by panghy — COMMENTED** [review]
**Review verdict: APPROVE** (submitted as a comment — GitHub does not allow approving one's own PR).

Reviewed all four commits with focus on path traversal/symlink escapes, TOCTOU, limit enforcement, error messages, tests, and Rust idioms. **No blocking issues found.** Seven non-blocking inline comments posted (TOCTOU hardening, deleted-root recreation edge case, data ToString coercion, file-count wording, check ordering, message consistency, poisoned-lock recovery in hot reload).

**Verified locally**: `cargo fmt --check` clean; all 18 `writeFile` sandbox unit tests, 11 config `write_dirs`/`resolve_write_roots` tests, 2 watcher hot-reload tests, and both `write_file_integration` tests pass. CI was still pending at review time.

**Security assessment**:
- The canonicalize-deepest-existing-ancestor + component-wise `starts_with` design is sound: symlinks in any existing part of the path are resolved before the prefix check, `..` is rejected outright, and the sibling-prefix string trick is covered by a test.
- A useful emergent property: a *dangling* symlink at the destination is treated as non-existing by the ancestor walk, and `rename(tmp, dest)` then **replaces the symlink instead of following it** — so it cannot be used to redirect a write. A live symlink at the destination is resolved by canonicalize and containment-checked. Consider a small test pinning the dangling-symlink-at-dest behavior so a future refactor (e.g. switching from rename to direct write) does not silently regress it.

**Test coverage suggestions (non-blocking)**: dangling-symlink-at-dest (above); destination-exists-as-directory (rename fails — error path is fine but untested).

**Question (non-blocking)**: the rejection message points users at "Settings → Write directories" in the desktop app — is that Settings section already shipped or planned in a companion desktop PR? If it lands later the message is fine as a forward-reference; just flagging so it does not dangle.

File: src/js_sandbox.rs (hunk 1 above)
Comment (panghy) [reply src/js_sandbox.rs:1213]: Implemented in c4d7b42 using the suggested belt-and-suspenders approach: resolve_write_path now returns the matched root, and write_atomic re-canonicalizes the destination directory after create_dir_all and re-asserts containment before writing.

File: src/js_sandbox.rs (hunk 2 above)
Comment (panghy) [reply src/js_sandbox.rs]: Documented in c4d7b42: write_atomic's doc comment now notes that a root deleted after config resolution is recreated by create_dir_all (still inside the allowed prefix) and that the never-auto-created stance in resolve_write_roots only covers resolution time.

File: src/js_sandbox.rs [hunk 8]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+                limits.max_total_bytes_per_run,
+                bytes_written
+            ))
+            .into());
+    }
+
+    let dest = resolve_write_path(&path_str, &write_roots)
+        .map_err(|msg| JsNativeError::error().with_message(msg))?;
+
+    write_atomic(&dest, &bytes).map_err(|msg| JsNativeError::error().with_message(msg))?;
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/js_sandbox.rs]: Partially implemented in c4d7b42: write_atomic now re-canonicalizes the destination directory after create_dir_all and re-asserts containment against the matched root, closing the practical validate→write window. Full openat2/no-follow handle-based traversal is platform-specific and out of scope here — a hostile local process racing the relay is outside the sandbox threat model (writeFile itself can only create regular files, so scripts cannot plant symlinks).

File: src/js_sandbox.rs [hunk 9]
```diff
@@ -824,6 +926,294 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+
+    // Per-file size cap, checked before base64 decoding so an oversized
+    // payload is rejected without allocating the decoded bytes. Base64
+    // padding means the ≈3n/4 estimate is an upper bound on the decoded
+    // size, so a payload that passes here cannot exceed the cap once decoded.
+    let estimated_len = match encoding.as_str() {
+        "utf8" => Some(data.len()),
+        "base64" => Some(data.len().saturating_mul(3) / 4),
+        _ => None,
+    };
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/js_sandbox.rs:1036]: Implemented in c4d7b42: the estimate now subtracts up to two trailing '=' padding bytes, and a regression test covers a payload whose decoded length is exactly the cap.

File: src/config.rs [hunk 10]
```diff
@@ -328,6 +338,35 @@ pub fn validate_profiles(config: &Config) -> Result<(), Vec<String>> {
...
+pub fn validate_write_dirs(config: &Config) -> Result<(), Vec<String>> {
+    let dirs = match &config.relay.write_dirs {
+        Some(d) if !d.is_empty() => d,
+        _ => return Ok(()),
+    };
+
+    let errors: Vec<String> = dirs
+        .iter()
+        .filter(|d| !expand_tilde(d).is_absolute())
+        .map(|d| {
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/config.rs:355]: Implemented in c4d7b42: expand_tilde now falls back to dirs::home_dir() when HOME is unset. HOME still takes precedence so tests that override it keep working (dirs::home_dir reads HOME on Unix anyway).

File: tests/fixtures/image_mcp_server.sh [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+# { type: "image", mimeType: "image/jpeg", data: <base64> } carrying a
+# synthetic 322-byte JPEG (SOI FF D8 FF ... EOI FF D9). The same base64
+# payload is hardcoded as JPEG_BASE64 in tests/write_file_integration.rs.
+
+IMG_B64="/9j/4AAQSkZJRgABAQAAAQABAAADChEYHyYtNDtCSVBXXmVsc3qBiI+WnaSrsrnAx87V3OPq8fj/Bg0UGyIpMDc+RUxTWmFob3Z9hIuSmaCnrrW8w8rR2N/m7fT7AgkQFx4lLDM6QUhPVl1ka3J5gIeOlZyjqrG4v8bN1Nvi6fD3/gUMExohKC82PURLUllgZ251fIOKkZifpq20u8LJ0Nfe5ezz+gEIDxYdJCsyOUBHTlVcY2pxeH+GjZSboqmwt77FzNPa4ejv9v0ECxIZICcuNTxDSlFYX2ZtdHuCiZCXnqWss7rByM/W3eTr8vkABw4VHCMqMTg/Rk1UW2JpcHd+hYyTmqGor7a9xMvS2eDn7vX8AwoRGB8mLTQ7QklQV15lbHN6gYiPlp2kq7K5wMfO1dzj6vH4/wYNFBsiKTD/2Q=="
+
+while IFS= read -r line; do
+    # Parse the method from JSON (simple grep-based extraction)
+    method=$(echo "$line" | sed -n 's/.*"method"[[:space:]]*:[[:space:]]*"\([^"]*\)".*/\1/p')
+    id=$(echo "$line" | sed -n 's/.*"id"[[:space:]]*:[[:space:]]*\([0-9]*\).*/\1/p')
```
[the remark this answers is not in the document]
Comment (panghy) [reply tests/fixtures/image_mcp_server.sh:15]: Implemented in c4d7b42: the fixture now skips the response entirely when the request has no id, so notifications no longer produce the malformed '{"id":}' line.

File: src/js_sandbox.rs (hunk 3 above)
Comment (panghy) [reply src/js_sandbox.rs]: Implemented in c4d7b42: non-string data is now rejected with TypeError 'writeFile: data must be a string', with a regression test.

File: src/js_sandbox.rs (hunk 4 above)
Comment (panghy) [reply src/js_sandbox.rs:1008]: Implemented in c4d7b42: the message now reads 'per-run limit of N file writes reached', matching what the counter actually counts.

File: src/js_sandbox.rs (hunk 5 above)
Comment (panghy) [reply src/js_sandbox.rs:1061]: Implemented in c4d7b42: resolve_write_path now runs before decoding, so path errors take precedence and disallowed paths never pay the decode allocation. Regression test added; the existing size-cap-before-decode test still passes (the estimate check remains first).

File: src/js_sandbox.rs (hunk 6 above)
Comment (panghy) [reply src/js_sandbox.rs:1175]: Implemented in c4d7b42: the rejection message now quotes the path like every other writeFile message.

File: src/watcher.rs (hunk 7 above)
Comment (panghy) [reply src/watcher.rs]: Implemented in c4d7b42: the hot-reload swap now uses write().unwrap_or_else(|p| p.into_inner()) so a poisoned lock degrades instead of freezing the allowlist, mirroring the read side.

File: src/js_sandbox.rs [hunk 12]
```diff
@@ -824,6 +926,335 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+    ));
+    if let Err(e) = std::fs::write(&tmp, bytes) {
+        let _ = std::fs::remove_file(&tmp);
+        return Err(format!(
+            "writeFile: failed to write '{}': {}",
+            dest.display(),
+            e
+        ));
+    }
+    if let Err(e) = std::fs::rename(&tmp, &final_dest) {
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/js_sandbox.rs:1260]: Rust's std::fs::rename does replace an existing destination file on Windows: it is implemented via MoveFileExW with MOVEFILE_REPLACE_EXISTING (see the std::fs::rename docs — 'This will not work if the new name is on a different mount point' is the caveat; replacing an existing file is supported on Windows since Rust 1.0 for files, only directories cannot be replaced). writeFile destinations are always regular files, so the overwrite path is atomic on all three release targets. No change needed.

File: src/js_sandbox.rs [hunk 13]
```diff
@@ -824,6 +926,335 @@ fn register_json_parse_wrapper(context: &mut Context) -> Result<(), JsSandboxErr
...
+            "writeFile: '{}' escaped the configured write directory during the write",
+            dest.display()
+        ));
+    }
+    let final_dest = canonical_dir.join(file_name);
+    let unique = std::time::SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)
+        .map(|d| d.as_nanos())
+        .unwrap_or(0);
+    let tmp = canonical_dir.join(format!(
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/js_sandbox.rs:1238]: Implemented in 5dc0bbf: the temp name now includes a process-wide monotonic AtomicU64 counter in addition to pid + timestamp, and the temp file is opened with create_new(true) so each write exclusively owns its temp file even on a name collision across processes.
